src/seq/coin_flipper.rs

```diff
@@ -0,0 +1,138 @@
+        // If 2n < d, n * 2^c < d
+        // If the result is all heads, then set n to n * 2^c
+        // If there was at least one tail, return false
+        // If 2n >= d, the order of the heads and tails matters so we flip one coin at a time so c = 1
+        // Ideally, c will be as high as possible within these constraints
+
+        while n < d {
+            //Find a good value for c by counting leading zeros
+            //This will either give the highest possible c, or 1 less than that
+            let c = n.leading_zeros().saturating_sub(d.leading_zeros() + 1).min(32).max(1);
+
+            // set next_n to n * 2^c (checked_shl will fail if 2n >= `usize::max`)
+            if let Some(next_n) = n.checked_shl(c) {
+                if self.flip_until_tails(c) {
+                    //All heads
+                    //if 2n < d, set n to 2n
+                    //if 2n >= d, the while loop will exit and we will return `true`
+                    n = next_n
+                } else {
+                    //At least one tail - either return false or set n to 2n-d
+                    n = next_n.saturating_sub(d);
+                    
+                    if n == 0 {                        
+                        //Because we used saturating_sub, n will be zero if 2n was less than d or 2n was equal to d, in either case we can return false.
+                        return false;
+                    }
+                }
+            } else {
+                // This branch will only be reached when 2n >= `usize::max`
+                // Obviously 2n > d
+                if self.flip_until_tails(1) {
+                    //heads
+                    return true;
+                } else {
+                    //tails
+                    n = n.saturating_add(n).saturating_sub(d); // set n to 2n -d
+                }
+            }
+        }
+        true
+    }
+        
+    /// If the next `c` bits of randomness are all zeroes, consume them and return true.
+    /// Otherwise return false and consume the number of zeroes plus one
+    /// Generates new bits of randomness when necessary (int 32 bit chunks)
+    /// Has a one in 2 to the `c` chance of returning true
+    /// `c` must be less than or equal to 32
+    fn flip_until_tails(&mut self, mut c: u32) -> bool {
+        debug_assert!(c <= 32); //If `c` > 32 this wil always return false
+                                //Note that zeros on the left of the chunk represent heads. It needs to be this way round because zeros are filled in when left shifting
+        loop {
+            let zeros = self.chunk.leading_zeros();
+
+            if zeros < c {
+                // The happy path - we found a 1 and can return false
+                // Note that because a 1 bit was detected, we cannot have run out of random bits so we don't need to check
+
+                // First consume all of the bits read
+                self.chunk = self.chunk.wrapping_shl(zeros + 1);
```

Comment:
Why wrapping shift instead of ordinary shift?

Reply:
I did it this way because, in principle `zeros` could be 31 which would cause a panic.
I tried changing it to ```self.chunk = self.chunk << zeros + 1;``` and changing the max of c to 31. This has a very strange effect on performance - The unhinted version gets about 20% faster but the size-hinted version (which shouldn't really be effected) got a lot shower (literally twice as slow with 1000 elements). I'm guessing this is because the compiler still thinks it could panic and so has to write extra code for that?